HIP: Enable Matrix cores for MMQ Kernels, Enable stream-K for CDNA 3

deepsek opened the pull request:
- Added Matrix cores support (MFMA instructions) for MMQ kernels. 
- Enable stream-K for CDNA3 to work with MMQ kernels.
- Removed usage of WARP_SIZE hardcoded constant in MMQ kernels.

- NOTE: Thoughts on removing all uses of hardcoded const specific to only NVIDIA (like WARP_SIZE) in order to support other GPUs?

@JohannesGaessler @ggerganov 
P.S. I am part of an AMD team actively working on enabling AMD feature set on llama.cpp. We would like to get on call to discuss some future PR plans for additional backends, flash attention changes, etc. 

EDIT:
Update to add some performance charts for DeepSeekV3 model.

_**Upstream vs ROCm Fork Development**_
<img width="1924" height="1080" alt="image" src="https://github.com/user-attachments/assets/2959803b-6050-4761-a4bc-7a700f636d0e" />

_**MI300X vs H100 Throughput Test**_
<img width="1921" height="1081" alt="image" src="https://github.com/user-attachments/assets/53a9d106-c027-4780-bda8-fe00fd1b5fe3" />

deepsek committed:
Feat: Enable MFMA instr for Q4_K

deepsek committed:
Fix: Missed template param

deepsek committed:
Feat: Add MFMA instr for Q6_K, remove MMQ_NWARPS

deepsek committed:
Merge branch 'ggml-org:master' into amd-integration

deepsek committed:
Merge branch 'ggml-org:master' into amd-integration

deepsek committed:
Merge branch 'ggml-org:master' into amd-integration

deepsek committed:
Perf: Fix Register Spilling Q6_K - Refactor kernel, launch_bound

deepsek committed:
Perf: Refactor Q4_K, reduce register pressure

deepsek committed:
Perf: Throughput Increase 4k->6.9k t/s

deepsek committed:
Perf: 7.1k tokens/sec

deepsek committed:
Perf/Feat: Throughput 8.3k tokens/sec, Add support for all quants

deepsek committed:
Feat: Remove warnings, deprecated __AMDGCN_WAVEFRONT_SIZE

deepsek committed:
Merge branch 'master' into amd-integration

deepsek committed:
Feat: Enable stream-k for CDNA3

ggerganov commented:
> P.S. I am part of an AMD team actively working on enabling AMD feature set on llama.cpp. We would like to get on call to discuss some future PR plans for additional backends, flash attention changes, etc.

@deepsek Thanks for the contribution and for reaching out. On topics related to the CUDA backend, @JohannesGaessler is the best person to consult with. For additional backends, @slaren can provide guidelines and advice. I'll be happy to provide input on any matters as well.

I am also available for call - feel free to contact me.

Dampfinchen commented:
Very nice to see the initiative. I assume improvements made for CDNA will also swap into the consumer side next year when UDNA releases.

JohannesGaessler commented:
Sorry, I wanted to ask: @IMbackK since you've been working on AMD support, are you interested in joining the discussion?

deepsek committed:
Fix: Remove Trailing Whitespaces

IMbackK commented:
> Sorry, I wanted to ask: @IMbackK since you've been working on AMD support, are you interested in joining the discussion?

Yes, certainly. It would help to avoid duplication of effort.

deepsek committed:
Fix: Unused Params Warnings, CUDA Build

deepsek committed:
-p512: 8.4k->9.5k - Account for DataPadding for writing tile_y

IMbackK commented:
@deepsek There are a few small things as discussed, better naming for this mfma path so that a rdna wmma solution can be added later without the nameing being strange is one thing, use of two V_MFMA_I32_16X16X16I8 instructions on gfx908 and gfx90a, even if this path is not chosen for those, to ease maintainability is another.

I would also like to try this myself on gfx94x somehow and i am not sure what the state is with regard to access to amds cloud for maintenance of a gfx94x specific code path, maybe @ggerganov can also comment on that. A problem here being that after cdna2/gfx90a/mi210 AMD has not made any further CDNA devices that are pcie addon board form factor, so out side of the acquisition of an entire mi300 oam machine no one can simply add a CDNA3/gfx94x/MI3xx compatible card to their system.

JohannesGaessler reviewed (commented):
This is only a partial review because I'm noticing that some calculations on master are wrong in terms of the logic (but correct in terms of the values). My suggestion for this PR is that I fix the incorrect logic on master first and that we then rebase this PR.

My code in `mma.cuh` implicitly assumes that for int8 and FP16 the raw data layout across threads in a warp is the same. Please tell me if this assumption is wrong for AMD.

JohannesGaessler commented:
>any suggestions on what you think would be appropriate? I see FP16_MMA_AVAILABLE being used. Thoughts on just using INT8_MMA_AVAILABLE?

If the instructions are only available on `CDNA3` my suggestion is `CDNA3_MMA_AVAILABLE`.

IMbackK commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -299,15 +447,15 @@ static __device__ __forceinline__ void vec_dot_q4_0_q8_1_dp4a(
 // #pragma unroll
-    for (int k01 = 0; k01 < WARP_SIZE; k01 += QR4_0*VDR_Q4_0_Q8_1_MMQ) {
+    for (int k01 = 0; k01 < MMQ_TILE_NE_K; k01 += QR4_0*VDR_Q4_0_Q8_1_MMQ) {
         const int k0 = k00 + k01;
 
 #pragma unroll
         for (int j0 = 0; j0 < mmq_x; j0 += nwarps) {
             const int j = j0 + threadIdx.y;
 
 #pragma unroll
-            for (int i0 = 0; i0 < mmq_y; i0 += WARP_SIZE) {
+            for (int i0 = 0; i0 < mmq_y; i0 += warp_size) {
```
Yes any GCN GPU

IMbackK commented:
> > any suggestions on what you think would be appropriate? I see FP16_MMA_AVAILABLE being used. Thoughts on just using INT8_MMA_AVAILABLE?
> 
> If the instructions are only available on `CDNA3` my suggestion is `CDNA3_MMA_AVAILABLE`.

the used MFMA instruction, V_MFMA_I32_16X16X32I8 is only available on CDNA3, the other CDNA devices support only k=16 ie V_MFMA_I32_16X16X16I8, obviously its pretty trivial to use V_MFMA_I32_16X16X32I8 vs V_MFMA_I32_16X16X16I8, whatever is available.

(consumer) rdna3+ gpus support the same-purpose V_WMMA_I32_16X16X16_IU8 but semantics are quite different.

I would suggest the obvious AMD_MFMA_AVAILABLE and AMD_WMMA_AVAILABLE with the AMD_MFMA_AVAILABLE path using V_MFMA_I32_16X16X32I8 or V_MFMA_I32_16X16X16I8 based on the presence of the CDNA3 define.

JohannesGaessler commented:
>This is only a partial review because I'm noticing that some calculations on master are wrong in terms of the logic (but correct in terms of the values). My suggestion for this PR is that I fix the incorrect logic on master first and that we then rebase this PR.

It seems I misremembered how the code works, the part I though was wrong is correct after all. I'll continue with the review.

deepsek commented on ggml/src/ggml-cuda/mma.cuh:
```diff
@@ -386,6 +441,46 @@ namespace ggml_cuda_mma {
+    static __device__ __forceinline__ void mma(
+            tile<32, 32, int> & D, const tile<32, 4, int> & A, const tile<32, 4, int> & B) {
+#if defined(AMD_MMA_AVAILABLE)
+#if defined(CDNA3)
+        using int32x16_t = __attribute__((__vector_size__(16 * sizeof(int)))) int;
+        int32x16_t* acc = (int32x16_t*) D.x;
+        acc[0] = __builtin_amdgcn_mfma_i32_32x32x16_i8(((int64_t*) A.x)[0],
+                                                       ((int64_t*) B.x)[0],
+                                                       acc[0],
+                                                       0, 0, 0);
+#elif defined(CDNA2) || defined(CDNA)
+#endif
```
Fixed

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -100,12 +100,12 @@ static int get_mmq_x_max_host(const int cc) {
 
 #if defined(GGML_USE_HIP) && defined(__HIP_PLATFORM_AMD__)
-    return 128;
+    return 64;
```
Yep. noticed during dev and fixed.

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
-#endif // NEW_MMA_AVAILABLE
+#endif // AMD_MMA_AVAILABLE
+
+static int get_mmq_nwarps_host(ggml_type type, const int cc) {
```
Also changed other functions like get_mmq_x_max_host, etc on main branch. (Replaced all get_mmq_* with mmq_get_*)

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
Are we sure we want to do this? Since Georgi also already test the performance and there was no regression. I'm just wondering, because even though this PR would generate different asm code, there is no change in performance on NVGPU. On the other hand, adding this conditional at every occurrence in the code makes it less clean IMO.

JohannesGaessler replied:
Georgi did not measure a difference on his GPU, that does not mean there will be no difference for all GPUs. Please just resolve this at compile time so I won't have to re-test all combinations of GPUs and data types.

JohannesGaessler replied:
If `nrows != 1` for both warp size 32 and 64 then no compile-time check is needed since it's not possible to optimize out instructions.

deepsek replied:
Resolved. Added wherever needed too

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -631,54 +808,109 @@ static __device__ __forceinline__ void vec_dot_q8_0_q8_1_dp4a(
+        }
+
+#pragma unroll
+        for (int j0 = 0; j0 < mmq_x; j0 += ntx*tile_C::J) {
```
I was wrong, this was correct - that is why I deleted the comment.

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -631,54 +808,109 @@ static __device__ __forceinline__ void vec_dot_q8_0_q8_1_dp4a(
+    constexpr int rows_per_warp = granularity;
+    constexpr int ntx = rows_per_warp/tile_C::I; // Number of x minitiles per warp.
+
+    y += (threadIdx.y % ntx) * (tile_C::J*MMQ_TILE_Y_K);
```
This was correct actually, sorry.

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1268,23 +1680,23 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
         }
 #else
-        x_sc[i*(WARP_SIZE/8) + i/8 + threadIdx.x % (WARP_SIZE/8)] = sc;
+        x_sc[i*4 + i/8 + ksc] = sc;
```
```suggestion
        x_sc[i*(MMQ_TILE_NE_K/8) + i/8 + ksc] = sc;
```

I think some semantic meaning is lost here. The reason for the hard-coded value 8 is that for every 8 rows with `WARP_SIZE` quant values there are `WARP_SIZE` scale values so you need 1 value of padding per 8 rows.

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
 
-    const int kbx  = threadIdx.x / QI4_0;
-    const int kqsx = threadIdx.x % QI4_0;
+    constexpr int threads_per_row = MMQ_ITER_K / (4 * QR4_0);
```
Ultimately I currently can't think of a solution that really makes me happy. I think for this PR we can just merge it as-is and if I can later think of a better way to do it we can refactor it then (provided you'll be around for testing since I don't have access to CDNA3 hardware).

deepsek replied:
I will be available. llama.cpp support on AMD compute is one of my teams primary focus. I will be happy to be involved in design discussions, testing, etc.

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1340,72 +1754,86 @@ static __device__ __forceinline__ int unpack_scales_q45_K(const int * scales, co
+        int i = i0 + threadIdx.y*rows_per_warp + threadIdx.x/2;
+        if (i < mmq_y) {
+#else
+        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/2) % mmq_y;
```
```suggestion
        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/2) % mmq_y;
        {
```

This way the indentation is consistent and you can add the `}` further down unconditionally.

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1415,30 +1843,32 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
         const int * scales = (const int *) bxi->scales;
 
-        const int ksc = threadIdx.x % (WARP_SIZE/8);
+        const int ksc = threadIdx.x % 4;
```
```suggestion
        const int ksc = threadIdx.x % (MMQ_TILE_NE_K/8);
```

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1415,30 +1843,32 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps * 8) {
-        int i = (i0 + threadIdx.y * 8 + threadIdx.x / (WARP_SIZE/8)) % mmq_y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nwarps*rows_per_warp) {
+        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/4) % mmq_y;
```
```suggestion
        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/(MMQ_TILE_NE_K/8)) % mmq_y;
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1415,30 +1843,32 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
         }
 
-        const block_q4_K * bxi = (const block_q4_K *) x + kbx0 + i*stride + (threadIdx.x % (WARP_SIZE/8)) / (QI4_K/8);
+        const block_q4_K * bxi = (const block_q4_K *) x + kbx0 + i*stride + (threadIdx.x % 4) / (QI4_K/8);
```
```suggestion
        const block_q4_K * bxi = (const block_q4_K *) x + kbx0 + i*stride + (threadIdx.x % (MMQ_TILE_NE_K/8)) / (QI4_K/8);
```

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1448,105 +1878,119 @@ static __device__ __forceinline__ void vec_dot_q4_K_q8_1_dp4a(
+        int i = i0 + threadIdx.y*rows_per_warp + threadIdx.x/2;
+        if (i < mmq_y) {
+#else
+        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/2) % mmq_y;
```
```suggestion
        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/2) % mmq_y;
        {
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1557,9 +2001,10 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps*8) {
-        int i = (i0 + threadIdx.y*8 + threadIdx.x/(WARP_SIZE/8)) % mmq_y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nwarps*rows_per_warp) {
+        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/4) % mmq_y;
```
```suggestion
        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/(MMQ_TILE_NE_K/8)) % mmq_y;
```

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1569,17 +2014,19 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
         const int * scales = (const int *) bxi->scales;
 
-        const int ksc = threadIdx.x % (WARP_SIZE/8);
+        const int ksc = threadIdx.x % 4;
```
```suggestion
        const int ksc = threadIdx.x % (MMQ_TILE_NE_K/8);
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1569,17 +2014,19 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
         const int scales8 = unpack_scales_q45_K(scales, ksc);
 
-        x_sc[i*(WARP_SIZE/8) + i/8 + ksc] = scales8;
+        x_sc[i*4 + i/8 + ksc] = scales8;
```
```suggestion
        x_sc[i*(MMQ_TILE_NE_K/8) + i/8 + ksc] = scales8;
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1589,112 +2036,118 @@ static __device__ __forceinline__ void vec_dot_q5_K_q8_1_dp4a(
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps * 8) {
-        int i = (i0 + threadIdx.y * 8 + threadIdx.x / (WARP_SIZE/8)) % mmq_y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nwarps*rows_per_warp) {
+        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/4) % mmq_y;
```
```suggestion
        int i = (i0 + threadIdx.y*rows_per_warp + threadIdx.x/(MMQ_TILE_NE_K/8)) % mmq_y;
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1589,112 +2036,118 @@ static __device__ __forceinline__ void vec_dot_q5_K_q8_1_dp4a(
         }
 
-        const block_q6_K * bxi = (const block_q6_K *) x + kbx0 + i*stride + (threadIdx.x % (WARP_SIZE/8)) / 4;
+        const block_q6_K * bxi = (const block_q6_K *) x + kbx0 + i*stride + (threadIdx.x % 4) / 4;
```
```suggestion
        const block_q6_K * bxi = (const block_q6_K *) x + kbx0 + i*stride + (threadIdx.x % (MMQ_TILE_NE_K/8)) / 4;
```

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1589,112 +2036,118 @@ static __device__ __forceinline__ void vec_dot_q5_K_q8_1_dp4a(
-#ifdef NEW_MMA_AVAILABLE
-        x_sc[i*MMQ_MMA_TILE_X_K_Q6_K + threadIdx.x % (WARP_SIZE/8)] = get_int_b2(bxi->scales, threadIdx.x % (QI6_K/8));
+#if defined(AMD_MMA_AVAILABLE) || defined(NEW_MMA_AVAILABLE)
+        x_sc[i*MMQ_MMA_TILE_X_K_Q6_K + threadIdx.x%4] = get_int_b2(bxi->scales, threadIdx.x%4);
```
```suggestion
        x_sc[i*MMQ_MMA_TILE_X_K_Q6_K + threadIdx.x%4] = get_int_b2(bxi->scales, threadIdx.x % (MMQ_TILE_NE_K/8));
```

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1589,112 +2036,118 @@ static __device__ __forceinline__ void vec_dot_q5_K_q8_1_dp4a(
+        x_sc[i*MMQ_MMA_TILE_X_K_Q6_K + threadIdx.x%4] = get_int_b2(bxi->scales, threadIdx.x%4);
 #else
-        x_sc[i*(WARP_SIZE/8) + i/8   + threadIdx.x % (WARP_SIZE/8)] = get_int_b2(bxi->scales, threadIdx.x % (QI6_K/8));
+        x_sc[i*4 + i/8 + threadIdx.x%4]               = get_int_b2(bxi->scales, threadIdx.x%4);
```
```suggestion
        // MMQ_TILE_NE_K == QI6_K
        x_sc[i*(MMQ_TILE_NE_K/8) + i/8 + threadIdx.x % (MMQ_TILE_NE_K/8)]               = get_int_b2(bxi->scales, threadIdx.x % (QI6_K/8));
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1862,54 +2370,60 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
+        x_df[i*MMQ_MMA_TILE_X_K_Q8_0        + kbxd] = __half2float(bxi->d);
 #else
-        x_df[i*(WARP_SIZE/4) + i/4   + kbxd] = __half2float(bxi->d);
+        x_df[i*(MMQ_TILE_NE_K/4) + i/QI4_NL + kbxd] = __half2float(bxi->d);
```
```suggestion
        x_df[i*(MMQ_TILE_NE_K/QI4_NL) + i/QI4_NL + kbxd] = __half2float(bxi->d);
```

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -2225,66 +2759,71 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+       int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
```suggestion
        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```

Also please add a compile-time check.

deepsek replied:
Resolved

JohannesGaessler commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -2225,66 +2759,71 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps * 4) {
-        int i = i0 + threadIdx.y * 4 + threadIdx.x / (WARP_SIZE/4);
+    for (int i0 = 0; i0 < mmq_y; i0 += nwarps * rows_per_warp) {
+        int i = i0 + threadIdx.y * rows_per_warp + threadIdx.x / 8;
```
```suggestion
        int i = i0 + threadIdx.y * rows_per_warp + threadIdx.x / (MMQ_TILE_NE_K/4);
```

deepsek replied:
Resolved

deepsek committed:
refactor: PR code cleanup, amd_mma_available->amd_mfma_available

deepsek commented on ggml/src/ggml-cuda/mma.cuh:
```diff
@@ -186,7 +220,11 @@ namespace ggml_cuda_mma {
+        int64_t* xi = (int64_t*) t.x;
+        const int64_t* xs = (int64_t*) ((const int*) xs0 + (threadIdx.x % t.I) * stride + 2 * (threadIdx.x / t.I));
+        xi[0] = xs[0];
+#elif defined(NEW_MMA_AVAILABLE)
```
Resolved. AMD MFMA tiles only use load_generic now.

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
+        }
+    } else if (new_mma_available(cc) && mmq_x >= 48) {
+        return 16;
+    } else return 8;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
+        case GGML_TYPE_Q6_K:   // vec_dot_q6_K_q8_1_mma
+        case GGML_TYPE_IQ2_XS: // vec_dot_q8_0_16_q8_1_mma
+        case GGML_TYPE_IQ2_S:  // vec_dot_q8_0_16_q8_1_mma
+            return mmq_x >= 192 ? 64 : 32;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
+            case GGML_TYPE_IQ3_S:
+            case GGML_TYPE_IQ4_XS:
+            case GGML_TYPE_IQ4_NL:
+                return 8;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
+        case GGML_TYPE_IQ3_S:
+        case GGML_TYPE_IQ4_XS:
+        case GGML_TYPE_IQ4_NL:
+            return 8;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
+}
+
+#if defined(AMD_MMA_AVAILABLE)
+static constexpr __device__ int get_mmq_nwarps_device(ggml_type type) {
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -215,42 +220,182 @@ static constexpr __host__ __device__ int mmq_get_mma_tile_x_k(ggml_type type) {
-    const int kqsx = threadIdx.x % QI4_0;
+    constexpr int threads_per_row = MMQ_ITER_K / (4 * QR4_0);
+    constexpr int nrows = warp_size / threads_per_row;
+    const int txi = threadIdx.x % threads_per_row;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -320,32 +468,37 @@ static __device__ __forceinline__ void vec_dot_q4_0_q8_1_dp4a(
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -415,32 +571,37 @@ static __device__ __forceinline__ void vec_dot_q4_1_q8_1_dp4a(
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -465,54 +626,60 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -535,94 +702,104 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1340,72 +1754,86 @@ static __device__ __forceinline__ int unpack_scales_q45_K(const int * scales, co
-    for (int i0 = 0; i0 < mmq_y; i0 += nwarps) {
-        int i = i0 + threadIdx.y;
+    for (int i0 = 0; i0 < mmq_y; i0 += nrows*nwarps) {
+        int i = i0 + threadIdx.y*nrows + threadIdx.x/threads_per_row;
```
Resolved

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1415,30 +1843,32 @@ template <int mmq_y, int nwarps, bool need_check> static __device__ __forceinlin
         const int scales8 = unpack_scales_q45_K(scales, ksc);
 
-        x_sc[i*(WARP_SIZE/8) + i/8 + ksc] = scales8;
+        x_sc[i*4 + i/8 + ksc] = scales8;
```
Resolved

deepsek commented:
@JohannesGaessler, I've refactored the code per your suggestions in the comments above.
@ggerganov, No regressions on my end too. All tests passed!
@IMbackK, Per your suggestion, renamed amd_mma to amd_mfma

IMbackK commented:
> @JohannesGaessler, I've refactored the code per your suggestions in the comments above. @ggerganov, No regressions on my end too. All tests passed!
> 
> @IMbackK,
> 
>     * Per your suggestion, renamed `amd_mma` to `amd_mfma`

Looks good, thank you.

>     * I've also added the MFMA instructions for CDNA2 (CDNA uses same instructions) in the `mma.cuh` file. Even though the instructions are added, the code path is disabled in `common.cuh` with `amd_mfma_available`. Same goes for `stream-K` in `mmq.cu` file.
> 
>     * I've tested for functionality, and it works. But its disabled by default because, for llama3.1-8B model on CDNA2, with `stream-K` and `MFMA` enabled, the performance is **better** only up until prompt size of 256/512. Anything above (512-4096 tested), the performance drops with the use of the matrix cores. I haven't had the chance to test on CDNA but I suspect it will be the same as CDNA2.

Thats good, im mainly concerned with being able to test this code path, for which this will do fine.

>     * We can maybe in the future, look into enabling MFMA code path for only smaller prompt sizes in the legacy CDNAs. But it's not in my roadmap at the moment.

I will take some time hopefully soon to tune for cdna1, thats what i mainly do around here after all.

JohannesGaessler reviewed (commented):
We could maybe use a single macro like `MMQ_USE_MMA` instead of `defined(AMD_MFMA_AVAILABLE) || defined(NEW_MMA_AVAILABLE)`. But I'm not sure whether that would make the code more legible or not (this should either way be easy to change with simple find and replace).

We should check whether using the larger warp size of 64 is better with the dp4a code (I would assume it is). Though if we can now use tensor cores with CDNA1/2 as well I think this is fairly low priority and fine to defer to a follow-up PR.

After my last round of comments has been addressed I would consider this PR fine to merge from my end.

IMbackK commented:
With regards to CDNA1/gfx908 performance i can not confirm the below:

>     * I've tested for functionality, and it works. But its disabled by default because, for llama3.1-8B model on CDNA2, with `stream-K` and `MFMA` enabled, the performance is **better** only up until prompt size of 256/512. Anything above (512-4096 tested), the performance drops with the use of the matrix cores. I haven't had the chance to test on CDNA but I suspect it will be the same as CDNA2.

rocblas code path:

```
| model                          |       size |     params | backend    | ngl | n_ubatch |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | -------: | --------------: | -------------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |      256 |          pp4096 |       1399.78 ± 1.64 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |      512 |          pp4096 |      1722.61 ± 92.06 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |     1024 |          pp4096 |      2036.20 ± 80.62 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |     2048 |          pp4096 |      1729.29 ± 24.48 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |     4096 |          pp4096 |       1722.21 ± 4.95 |
```

this prs code path enabled on gfx908:

```
| model                          |       size |     params | backend    | ngl | n_ubatch |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | -------: | --------------: | -------------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |      256 |          pp4096 |       2071.35 ± 4.10 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |      512 |          pp4096 |       2125.70 ± 5.57 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |     1024 |          pp4096 |       2212.21 ± 4.57 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |     2048 |          pp4096 |      2114.55 ± 20.10 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | ROCm       |  99 |     4096 |          pp4096 |       2073.73 ± 2.14 |
```

This is pretty strange, given how similar gfx908 and gfx90a are and suggests that either something is not ideal with code generation for gfx90a or tensile kernels are poorly tuned for gfx908, perhaps due to gfx90a's occupancy advantage when a kernel allocates less accumulation registers than arch registers, but even this seams unlikely in a kernel prominently using mfma.

Either way, it would seam that gfx908 should be enabled to take this code path, you can add this to this pr or i will open another one right after, either is fine.

IMbackK commented:
> - I'll add this code path for all CDNA architectures. Should be as simple as enabling the `amd_mfma_available` option.
> - @IMbackK, did you also test with stream-K enabled for CDNA? You would have to change it in `mmq.cu` file as a 
`use_stream_K` flag and also in `mmq.cuh`, the code path in `mul_mat_q` kernel. (Check for CDNA3 and change it to CDNA).
> - If you can confirm stream-K with CDNA too, I can edit that in this same PR!

This was the full code path of this pr, mfma and stream-K. However this path introduces rather severe regressions for some datatypes in the op test (eg q2_K)

> - Regarding the edit, the variability could arise from the fact that the code is different quants, with different tile sizes, etc. For gfx942, i didn't test the tflops variability. I did measure overall throughput increase across all quantization types and several models. 

While the average goes up, on gfx908 the path the pr introduces appears to introduce quite large regressions for some datatypes, eg q2_K, i will download a corresponding model and do a more realistic test, but a severe regression like that in some datatypes is not really acceptable, please check if gfx942 is also affected.

> - Are you talking about the TFLOPS? Because I see an increase on average across various quants. 
> - Also, can you clarify what the difference between the `main branch dp4a code path` and `main branch code path` from the previous comment?  

Yes while some datatypes improve in performance several like q4_0, q4_1, and q4_k encounter mild to severe regressions in test-backend-ops perf, i will do some tests with corresponding models.

"main branch dp4a code path" refers to the mmq code path with no usage of mfma but with usage of __builtin_amdgcn_sdot4. This is the code path a Vega device takes see `ggml_cuda_should_use_mmq`.

"main branch code path" is the dequant -> rocblas code path taken by CDNA and RDNA3+ devices on the master branch again see `ggml_cuda_should_use_mmq`.

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -849,45 +1131,93 @@ static __device__ __forceinline__ void vec_dot_q8_0_16_q8_1_dp4a(
+    constexpr int rows_per_warp = granularity;
+    constexpr int ntx = rows_per_warp/tile_C::I; // Number of x minitiles per warp.
+
+    y += (threadIdx.y % ntx) * (tile_B::I*MMQ_TILE_Y_K);
```
Changing instances of `tile_C::J` to `tile_B::I` instead due to lesser occurrence (convenience in editing). Waiting for confirmation before pushing commit

deepsek replied:
Switched to `tile_C::J`

deepsek commented on ggml/src/ggml-cuda/mmq.cuh:
```diff
@@ -1052,45 +1387,110 @@ static __device__ __forceinline__ void vec_dot_q2_K_q8_1_dp4a(
+    constexpr int rows_per_warp = granularity;
+    constexpr int ntx = rows_per_warp/tile_C::I; // Number of x minitiles per warp.
+
+    y += (threadIdx.y % ntx) * (tile_B::I*MMQ_TILE_Y_K);
```
Changing instances of `tile_C::J` to `tile_B::I` instead due to lesser occurrence (convenience in editing). Waiting for confirmation before pushing commit

JohannesGaessler commented:
For NVIDIA GPUs and AMD GPUs without `AMD_MFMA_AVAILABLE` a constant number of 8 warps is used. Previously this meant 256 threads but with the warp size now potentially being 64 it can mean 512 threads. I suspect there are then issues with register pressure. So I would suggest trying 4 warps of size 64 instead - my intuition is that with the same number of threads but organized to match the physical warp size the performance will be better than on master.

deepsek committed:
refactor: PR cleanup

IMbackK commented:
> So, if I'm understanding right. gfx908 shows same regression, as my initial tests with gfx90a with the MFMA instructions. Yea, I agree this shouldn't be a blocker for gfx942 (this PR). Tbh, I don't see enabling matrix cores on the older CDNA architectures in the near future (esp, cause MI100 is 'discontinued'). We can stick with the current code path on main branch for the older archs.

I certainly do, but this is not related to this pr

> My 2 cents without any further investigations. the rocblas/tensile kernels have not been tuned for a long while now. That could explain why MI100 is outperforming MI300. There is a push to migrate to hipBlasLt instead. This migration for llama.cpp is also on the roadmap for my team. But again, this is unrelated to this PR. We will open another PR for this matter down the road.

While llamacpp could use hipBlasLt, hipBlasLt is a clone of cublaslt and as such not a reasonable replacement actual blas, the suggestion that the existence of lt frees amd from the requirement to tune rocblas to a sensible level is frankly absurd, but i understand this is not your department.

If you do open a pr that adds support to dispatch to hipblaslt, understand that we can not accept this replacing hipblas/rocblas, as the list of devices supported by hipblaslt is absolutely tiny, we have users that use devies as old as gfx803 successfully while hipblaslt dosent even support the cdna architecture from 2020, never mind older and or consumer devices.

>     * On CDNA3, for MFMA PR code path, q2_K shows regression from `83.39 TFLOPS` to `71.49 TFLOPS`. While other quants jump from `~80 TFlOPS` to `~200 TFLOPS`.

I find this fine to ignore , if you want, given the large improvement to the other types.
 
>     * On GCN (vega20), for the dp4a code path, Q4_0 and Q4_1 show regression from `~15 TFLOPS` to `~10 TFLOPS`. While all other quants show an increase. (@0cc4m)

This one i find unacceptable.

Further i think it needs to be examined if this pr is necessary at all given that it would seam that the current code path taken (dequant->rocblas) could perform better than this pr if run with `export ROCBLAS_USE_HIPBLASLT=1` which causes rocblas to cross dispatch to hipblaslt or if bf16 is used as a dequant target.

IMbackK commented:
Unfortunately i cant test the last one myself as for now i was denied time on the amd developer cloud

IMbackK commented:
> > Further i think it needs to be examined if this pr is necessary at all given that it would seam that the current code path taken (dequant->rocblas) could perform better than this pr if run with `export ROCBLAS_USE_HIPBLASLT=1` which causes rocblas to cross dispatch to hipblaslt or if bf16 is used as a dequant target.
> 
> I'm leaning to agree with @JohannesGaessler on this. While the prompt size 512+ for q2_k is lower right now, that should be fixed soon. But we can lean to dp4a, if I'm not able to resolve it soon.

The low performance with q2_k is mainly vs rocblas/hipblaslt not dp4a as dp4a is only used for small batch sizes and the issue is mainly at large batch sizes, but yes we can just use the rocblas/hipblasl code path for this datatype.

Your benchmark results suggest that like CDNA1/2 with rocblas, on CDNA3 with hipblaslt the performance for large batch sizes varies with datatype and sometimes hipblaslt wins and sometimes this pr wins. This is not a big problem, but ideally this pr would also change ggml_cuda_should_use_mmq to limit its own applicability to the fast datatypes. Since this is now the same across all CDNA devices the code path should ideally be the same across all CDNA devices. But since both hipblaslt and rocblas are currently also broken on CDNA3 we might want to skip the blas path on CDNA3 unconditionally, as this pr dose.

Please lean on your BLAS colleagues to have rocblas cross dispatch to hipblaslt automatically on GFX942 where this is faster, as it already dose on GFX12

So to recap these are the steps to take in my opinion:

1. change ggml_cuda_should_use_mmq to return false for q2_k
2. hopefully fix the gfx906 regression by trying nwarps = 4
3. merge this pr
4. flow up pr enables this pr for gfx908/gfx90a on datatypes where it outperforms rocblas
    - i will do this
6. fix the hipblaslt bug
7. limit the applicability of this pr to the same fast datatypes on gfx942 as on gfx908/90a from step 4 and use hipblaslt on gfx942

deepsek committed:
Feat/Perf: redesign all quants use same tile size, mfma instr, nwarps

IMbackK commented:
> @JohannesGaessler @IMbackK
>       * I have requested access to a Vega20 system, but it might get delayed. @0cc4m, would appreciate it if you can run this test on your setup. Also, gfx906 support is discontinued in rocm 6.4.0. Which rocm version are you using?

Since CDNA is mostly just GCN renamed with mfma added and the gfx pipeline remove, a pretty good facsimile of gfx906 can be constructed by forcing a CDNA device to take the dp4a path, we can add a compile time option similar to GGML_CUDA_FORCE_MMQ to make this happen, to allow you to more easly check if your changes are negatively affecting GCN.

Generally in the llamacpp community distro packages, as those provided by arch linux, are most popular. These provide a greatly superior expirance to amds official install method and still support gfx906 and older and will likely continue to do so for as long as possible.

>     * **_More importantly_**, I tested the main branch on a H100, With the default code path (MMQ with mma instr) and with `ggml_cuda_should_use_mmq` returning false.

The code is mainly targeted at consumer devices. I dont think anyone has ever tuned llamacpp/ggml for H100.
It also makes sense that cublas is more likely to win on H100 than consumer devices, as presumably NV spends more time tuning its blas libraries for these devices.
 
Id also like to note here that even in hipblaslt, it seams that that the fp32 result on mi300 is to way to low and is something someone from the blas libaries team should investigate. I get that since CDNA3 has 304 CUs it can be hard to fill all the cus due to GCN/CDNA's 4x smid16 nature requiring alot of threads to fill, but mi300 barely outperforming MI100 despite haveing over twice the CUs seams wrong.



If your performance assertions with regards to this pr prove true and the performance on gfx906 has be de-regressed, i will be happy to merge this pr.

deepsek committed:
Fix: CI fail - unused parameter Werror

deepsek commented:
Fixed the unused parameter Werror.

> The code is mainly targeted at consumer devices. I dont think anyone has ever tuned llamacpp/ggml for H100.
> It also makes sense that cublas is more likely to win on H100 than consumer devices, as presumably NV spends more time tuning its blas libraries for these devices.

I quickly ran the test on a RTX 2080 ti. I see the same results with cublas path outperforming the MMQ path, same as H100.
Again, while this is not a blocker for this PR. 
I'm looking to see if there a reasoning why a different design choice is made here between AMD & NV in terms of code paths, or if I'm just doing something wrong with my testing. Or maybe I'm just overthinking here and it slipped through the cracks during testing and needs to be fixed? 

```
MMA Disabled (cublas path)
  MUL_MAT(type_a=f32,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):               206 runs -  4890.33 us/run -  60.13 GFLOP/run -  12.30 TFLOPS
  MUL_MAT(type_a=f16,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):              1232 runs -   812.32 us/run -  60.13 GFLOP/run -  74.02 TFLOPS
  MUL_MAT(type_a=bf16,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      130 runs -  7757.88 us/run -  60.13 GFLOP/run -   7.75 TFLOPS
  MUL_MAT(type_a=q4_0,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      926 runs -  1081.20 us/run -  60.13 GFLOP/run -  55.61 TFLOPS
  MUL_MAT(type_a=q4_1,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      930 runs -  1075.81 us/run -  60.13 GFLOP/run -  55.89 TFLOPS
  MUL_MAT(type_a=q5_0,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      868 runs -  1154.16 us/run -  60.13 GFLOP/run -  52.10 TFLOPS
  MUL_MAT(type_a=q5_1,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      870 runs -  1150.04 us/run -  60.13 GFLOP/run -  52.28 TFLOPS
  MUL_MAT(type_a=q8_0,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      898 runs -  1113.99 us/run -  60.13 GFLOP/run -  53.98 TFLOPS
  MUL_MAT(type_a=q2_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      950 runs -  1053.73 us/run -  60.13 GFLOP/run -  57.06 TFLOPS
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      932 runs -  1075.19 us/run -  60.13 GFLOP/run -  55.92 TFLOPS
  MUL_MAT(type_a=q4_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      936 runs -  1069.79 us/run -  60.13 GFLOP/run -  56.21 TFLOPS
  MUL_MAT(type_a=q5_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      938 runs -  1067.14 us/run -  60.13 GFLOP/run -  56.35 TFLOPS
  MUL_MAT(type_a=q6_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      932 runs -  1074.55 us/run -  60.13 GFLOP/run -  55.96 TFLOPS
  MUL_MAT(type_a=iq2_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                   660 runs -  1517.49 us/run -  60.13 GFLOP/run -  39.62 TFLOPS
  MUL_MAT(type_a=iq2_xs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                    598 runs -  1674.09 us/run -  60.13 GFLOP/run -  35.92 TFLOPS
  MUL_MAT(type_a=iq2_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     650 runs -  1542.34 us/run -  60.13 GFLOP/run -  38.99 TFLOPS
  MUL_MAT(type_a=iq3_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                   658 runs -  1520.42 us/run -  60.13 GFLOP/run -  39.55 TFLOPS
  MUL_MAT(type_a=iq1_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     780 runs -  1283.44 us/run -  60.13 GFLOP/run -  46.85 TFLOPS
  MUL_MAT(type_a=iq1_m,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     616 runs -  1626.38 us/run -  60.13 GFLOP/run -  36.97 TFLOPS
  MUL_MAT(type_a=iq4_nl,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                    758 runs -  1319.47 us/run -  60.13 GFLOP/run -  45.57 TFLOPS
  MUL_MAT(type_a=iq3_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     696 runs -  1439.55 us/run -  60.13 GFLOP/run -  41.77 TFLOPS
  MUL_MAT(type_a=iq4_xs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                    784 runs -  1277.54 us/run -  60.13 GFLOP/run -  47.07 TFLOPS
  
MMA Enabled (Default)
  MUL_MAT(type_a=f32,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):               206 runs -  4860.04 us/run -  60.13 GFLOP/run -  12.37 TFLOPS
  MUL_MAT(type_a=f16,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):              1262 runs -   793.19 us/run -  60.13 GFLOP/run -  75.81 TFLOPS
  MUL_MAT(type_a=bf16,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      134 runs -  7536.22 us/run -  60.13 GFLOP/run -   7.98 TFLOPS
  MUL_MAT(type_a=q4_0,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      840 runs -  1191.10 us/run -  60.13 GFLOP/run -  50.48 TFLOPS
  MUL_MAT(type_a=q4_1,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      760 runs -  1317.36 us/run -  60.13 GFLOP/run -  45.64 TFLOPS
  MUL_MAT(type_a=q5_0,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      774 runs -  1294.64 us/run -  60.13 GFLOP/run -  46.44 TFLOPS
  MUL_MAT(type_a=q5_1,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      718 runs -  1395.47 us/run -  60.13 GFLOP/run -  43.09 TFLOPS
  MUL_MAT(type_a=q8_0,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      702 runs -  1425.90 us/run -  60.13 GFLOP/run -  42.17 TFLOPS
  MUL_MAT(type_a=q2_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      448 runs -  2237.44 us/run -  60.13 GFLOP/run -  26.87 TFLOPS
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      652 runs -  1535.98 us/run -  60.13 GFLOP/run -  39.15 TFLOPS
  MUL_MAT(type_a=q4_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      768 runs -  1303.45 us/run -  60.13 GFLOP/run -  46.13 TFLOPS
  MUL_MAT(type_a=q5_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      722 runs -  1388.57 us/run -  60.13 GFLOP/run -  43.30 TFLOPS
  MUL_MAT(type_a=q6_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                      634 runs -  1577.74 us/run -  60.13 GFLOP/run -  38.11 TFLOPS
  MUL_MAT(type_a=iq2_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                   622 runs -  1609.05 us/run -  60.13 GFLOP/run -  37.37 TFLOPS
  MUL_MAT(type_a=iq2_xs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                    412 runs -  2435.66 us/run -  60.13 GFLOP/run -  24.69 TFLOPS
  MUL_MAT(type_a=iq2_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     490 runs -  2047.03 us/run -  60.13 GFLOP/run -  29.37 TFLOPS
  MUL_MAT(type_a=iq3_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                   570 runs -  1756.19 us/run -  60.13 GFLOP/run -  34.24 TFLOPS
  MUL_MAT(type_a=iq1_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     490 runs -  2043.27 us/run -  60.13 GFLOP/run -  29.43 TFLOPS
  MUL_MAT(type_a=iq1_m,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     644 runs -  1554.53 us/run -  60.13 GFLOP/run -  38.68 TFLOPS
  MUL_MAT(type_a=iq4_nl,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                    654 runs -  1532.06 us/run -  60.13 GFLOP/run -  39.25 TFLOPS
  MUL_MAT(type_a=iq3_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                     626 runs -  1597.93 us/run -  60.13 GFLOP/run -  37.63 TFLOPS
  MUL_MAT(type_a=iq4_xs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=0):                    648 runs -  1545.26 us/run -  60.13 GFLOP/run -  38.91 TFLOPS
```

IMbackK commented:
> I quickly ran the test on a RTX 2080 ti. I see the same results with cublas path outperforming the MMQ path, same as H100. Again, while this is not a blocker for this PR. I'm looking to see if there a reasoning why a different design choice is made here between AMD & NV in terms of code paths, or if I'm just doing something wrong with my testing. Or maybe I'm just overthinking here and it slipped through the cracks during testing and needs to be fixed?

I only use amd devices so i cant comment too mutch on NV performance but one reason why the MMQ path is choosen on NV devices while amd deivces mostly choose the blas path is simply because @JohannesGaessler wrote most of the MMQ path while purely working on and optimizing for NV devices and I purely work on AMD devices and favored optimizing the blas path, as this was the path of less resistance.

deepsek commented:
@IMbackK, are we ready to merge on this PR or is there anything else pending?

Based on comments from @slaren and @IMbackK , looks like `ggml_cuda_should_use_mmq` needs to be modified to be more fine-tuned in choosing MMQ based on architectures and sizes.

@JohannesGaessler 
> To clarify, the function should return true for small batch sizes and false for large batch sizes with the boundary chosen in such a way that maximizes performance.

@slaren 
> The lower memory usage of MMQ is also a significant factor on why it is the default, even if it is not always faster than cuBLAS. The size of the buffer used to store the dequantized weights can be significant on consumer GPUs.

- Also is there a guidance as to when we prefer performance over memory usage? Per the 2 comments above, there seems to be multiple train of thoughts in design. 

P.S.
- Is there a documentation regarding these? Even while developing for MMQ, I am finding it hard to contribute, because many design choices and code is written implicitly without proper documentation/comments as to why the values/choices are chosen. 
- TBH, when adding support for newer architectures, it's hard for us to understand why the current values are sitting in the code without any explanation. It just takes way too long to figure these details out.

JohannesGaessler commented on ggml/src/ggml-cuda/mma.cuh:
```diff
@@ -148,10 +187,23 @@ namespace ggml_cuda_mma {
     template <int I, int J, typename T>
     static __device__ __forceinline__ void load_generic(tile<I, J, T> & t, const T * __restrict__ xs0, const int stride) {
+#if defined(AMD_MFMA_AVAILABLE)
+        if constexpr (I == 64 && J == 2) { // Special tile size to load <16, 4> as <16, 8>
+#pragma unroll
+            for (int l = 0; l < t.ne; ++l) {
+                t.x[l] = xs0[t.get_i(l)*stride + t.get_j(l)];
+            }
+        } else {
+            int64_t * xi = (int64_t *) t.x;
+            const int64_t * xs = (int64_t *) ((const int *) xs0 + (threadIdx.x % t.I) * stride + 2 * (threadIdx.x / t.I));
+            xi[0] = xs[0];
+        }
```
```suggestion
        if constexpr (I != 64 || J != 2) {
            int64_t * xi = (int64_t *) t.x;
            const int64_t * xs = (int64_t *) ((const int *) xs0 + (threadIdx.x % t.I) * stride + 2 * (threadIdx.x / t.I));
            xi[0] = xs[0];
            return;
        }
```

I think this would be simpler.

slaren commented:
> Is there actual guidance as to when performance is preferred over memory usage? Looks like there are conflicting viewpoints. Would be great to have this information documentation for when we contribute and add other architectures, there is a common design principle.

It's very hard to have a single design principle for every situation. Historically, some models would need as much as 6 GB of VRAM for the dequantization buffer. This would cause a lot of problems for people using consumer GPUs who would not understand why they could not offload to the GPU as many layers of the model as they would expect. This was one of the reasons why MMQ was made the default, even thought it was not faster than cuBLAS in every situation. Ultimately, it doesn't matter if cuBLAS is 10% faster, if using it means that you need to keep a large portion of the model in a CPU that is 10 times slower. For a data center GPU where VRAM is not so limited, the calculus may be different.

deepsek commented:
Sounds good. Great to hear the intuition process on these. 
My primary goal is to enable all possible feature set. I'll leave the code path, design choices and subsequent modifications to the `ggml_cuda_should_use_mmq` code for the community.
I am here if you need to run perf/bench, etc on the hardware that is openly available (primarily CDNA3, my focus atm).

JohannesGaessler commented:
I don't think the current solution is good but I'm willing to approve the PR as-is as long as you promise to refactor the code in the near future.

xbezdick commented:
Needed manual rebase but works great and on multiple AMD Radeon RX 7900 XT cards resolves "Page not present or supervisor privilege" gpu crash on k-shift.

IMbackK commented:
This pr should not affect gfx11 in any way, thats also the first i hear of this problem.

IMbackK reviewed (approved):
I have tested this pr one more time and am satisfied that there are no further regressions except the small regression in the dp4a code path in Q4_0 and Q4_1 on gfx9xx, which i find acceptable give the performance benefit in other areas.

JohannesGaessler commented:
I can reproduce the regression:

| Model                   | FlashAttention   | Test   |   t/s b5996 |   t/s b5997 |   Speedup |
|:------------------------|:-----------------|:-------|------------:|------------:|----------:|
| llama 8B Q4_0           | No               | pp512  |      786.04 |      556.47 |      0.71 |
| llama 8B Q4_0           | No               | tg128  |       62.46 |       62.51 |      1.00 |
| llama 8B Q4_0           | Yes              | pp512  |      604.58 |      459.13 |      0.76 |
| llama 8B Q4_0           | Yes              | tg128  |       59.81 |       59.91 |      1.00 |
| qwen3moe 30B.A3B Q3_K_S | No               | pp512  |      525.15 |      402.47 |      0.77 |
| qwen3moe 30B.A3B Q3_K_S | No               | tg128  |       39.97 |       39.98 |      1.00 |
| qwen3moe 30B.A3B Q3_K_S | Yes              | pp512  |      403.61 |      325.29 |      0.81 |
| qwen3moe 30B.A3B Q3_K_S | Yes              | tg128  |       37.85 |       37.91 |      1.00 |

I don't at all understand why this is happening though since the dp4a codepath for a warp size of 32 should be unchanged.

deepsek commented:
Good catch! Looks like the last line of the `ggml_use_mmq...` is using this code path for other amd gpus too. Missed this!

IMbackK commented:
@xbezdick since we have discovered that the RDNA path was affected by this path after all by accident, and we have restored original behavior with could you retest git and check if you can repoduce a crash there again and open an issue if so?
